website/blog/why-was-my-bazel-build-so-slow.md

```diff
@@ -0,0 +1,208 @@
+
+This post frequently references the BuildBuddy UI. While most of these steps are
+possible without it, it will be a lot more work to find and parse the info from Bazel yourself.
+If you’re not already using it, you can stream build events to our servers for free by
+adding the following flags to your builds (more setup docs [here](https://www.buildbuddy.io/docs/cloud)):
+
+`--bes_results_url=https://app.buildbuddy.io/invocation/`
+
+`--bes_backend=grpcs://remote.buildbuddy.io`
+
+Once you’ve added these flags, you’ll see a line like `Streaming build results to: https://app.buildbuddy.io/invocation/XXX` in your build logs. If you follow that link to our UI, you’ll have access to many of the features discussed below.
+
+## 1) Am I using the right set of flags?
+
+⭐ Bazel is very configurable with a huge selection of flags, so a good gut check is to verify which flags have been applied to your build. Here are a couple you may want to add (or verify have been set) to speed up your builds:
+
+- `--nolegacy_important_outputs`
+  - This flag can significantly reduce the payload size of the uploaded build event stream by eliminating duplicate file references.
+- `--remote_build_event_upload=minimal`
+  - This flag can significantly reduce network bandwidth by not uploading blobs tagged with no-cache to the cache.
+- `--remote_download_outputs`
+  - This flag can significantly reduce network bandwidth by not downloading intermediary build artifacts from the cache.
+- `--experimental_remote_cache_async`
+  - This flag can improve remote cache performance by enabling cache I/O in the background.
+- `--noremote_upload_local_results`
+  - This flag can reduce uploads by not uploading locally executed action outputs to the remote cache. This will reduce the cache hit rate for future runs, but can be desirable if upload speed is constrained (due to a poor network connection, for example).
+- `--remote_cache_compression`
+  - This flag can improve remote cache throughput by compressing cache blobs.
+- `--digest_function=BLAKE3`
+  - This flag can improve the performance of digest calculation of large files by using a faster hashing algorithm. This is available for Bazel 6.4+.
+- `--jobs`
+  - This flag can improve execution throughput by allowing more actions to execute in parallel.
+
+⭐ The details on the top of our UI are helpful for a quick sanity check here. If I’m comparing two builds, my first step will often be to compare the details on the top of the invocation pages and verify they’re exactly the same.
+
+For example, once I was inspecting a slow build, and realized I’d forgotten to set `--remote_download_minimal`. The build was downloading a lot more artifacts from the cache than usual, which explained the performance hit.
+
+![](../static/img/blog/ui-details.png)
+
+_The top invocation does not have –remote_download_minimal enabled, which could negatively impact performance._
```

Comment:
Maybe worth recommending these are all set via a `.bazelrc` that's checked into git?